### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by andmichalski**
When I override modeladmin - changelist_view on get_queryset method. How I can go inside this queryset in unittesting

Why in unittesting if I write self.c.post(....) when debuging and dig into response - method is GET.

In unittesting if I compare queryset from response.context_data (all children) with model (for example Child.objects.all() ) it is not equal.

Thanks

### CODE REVIEW

**Comment by andmichalski** [comment]
Add send_mail option and try using it with threading (beacuse threading module is using with socket. I want to train.) Mocked Thread not send_mail. Is it correct?

File: tree/admin.py [hunk 1]
```diff
@@ -92,16 +98,40 @@ def childlistview(self, request):
...
+        for i in range(num_fathers):
+            father = queryset[i]
+            email_address = father.email
+            child_list = [child.name + " " + child.last_name for child in father.child_set.all()]
+            text_message = "You have beautifull children: \n" + "\n".join(child_list) + "\nRegards"
+            t = Thread(target=send_mail, args=("Hello father",
+                                               text_message,
+                                               "admin@family.com",
+                                               email_address,))
+            t.start
```
Comment (rsiera) [inline tree/admin.py]: You don't even execute start ,it's a function t.start()

File: tree/admin.py [hunk 2]
```diff
@@ -92,16 +98,40 @@ def childlistview(self, request):
...
         return TemplateResponse(request, "admin/tree/father/parent_list.html", {'parents': parents})
 
+    def childs(self, obj):
+        return ", ".join([child.name for child in obj.child_set.all()])
+
+    def father(self, obj):
+        return obj
+
+    def send_email(self, request, queryset):
+        num_fathers = queryset.count()
```
Comment (rsiera) [inline tree/admin.py]: What ? You are getting count of queryset to make range on that, and then use this number to call __getitem__ on queryset ?

```
for father in queryset.prefetch_related('child_set').all():
    ...
```

File: tree/tests.py [hunk 3]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+    @patch('tree.admin.ChildAdmin.message_user')
+    def test_should_display_correct_message(self, mock_message):
+        url = reverse("admin:tree_child_changelist")
+        child = Child.objects.get(id=1)
+        data = {"action": "change_lastname", "_selected_action": [child.pk], "_last_name": "Pietraszek"}
+        response = self.client.post(url, data, follow=True)
+
+        self.assertEqual(mock_message.call_count, 1)
+
+        response.wsgi_request.method = "POST"
```
Comment (rsiera) [inline tree/tests.py]: What is that ?

File: tree/models.py [hunk 4]
```diff
@@ -15,13 +17,16 @@ def is_mine(self):
         qs = iam.child_set.all()
         return qs
 
+
 class Father(models.Model):
     name = models.CharField(max_length=50)
     last_name = models.CharField(max_length=50)
+    email = models.EmailField(null=True)
```
Comment (rsiera) [inline tree/models.py:24]: Where is migration for that ?

File: tree/tests.py [hunk 5]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+        data = {"action": "change_lastname", "_selected_action": [child.pk], "_last_name": "Pietraszek"}
+        response = self.client.post(url, data, follow=True)
+        self.assertEqual("Pietraszek", Child.objects.get(id=1).last_name)
+
+    @patch('tree.admin.ChildAdmin.message_user')
+    def test_should_display_correct_message(self, mock_message):
+        url = reverse("admin:tree_child_changelist")
+        child = Child.objects.get(id=1)
+        data = {"action": "change_lastname", "_selected_action": [child.pk], "_last_name": "Pietraszek"}
+        response = self.client.post(url, data, follow=True)
```
Comment (rsiera) [inline tree/tests.py]: Don't use follow=True, instead of changing method="POST" later.

File: tree/tests.py [hunk 6]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+
+        self.assertEqual(mock_message.call_count, 1)
+
+        response.wsgi_request.method = "POST"
+        mock_message.assert_has_calls(response.wsgi_request,
```
Comment (rsiera) [inline tree/tests.py]: better assert_called_once_with

File: tree/tests.py [hunk 7]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+                                      unicode("Changed in selected records last name to Pietraszek"))
+
+    def test_should_find_child_with_same_name(self):
+        father3 = Father.objects.create(name="Marcin", last_name="Tomasiak")
+        Child.objects.create(name="Jan", last_name="Tomasiak", birth=datetime(2018, 1, 5), father=father3)
+
+        url = reverse("admin:childlist")
+        data = {"name": "Jan"}
+        response = self.client.get(url, data, follow=True)
```
Comment (rsiera) [inline tree/tests.py]: What is follow=True for if you don't redirect anywhere in this view ?

File: tree/tests.py [hunk 8]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+
+        qs = response.context_data['childs']
+        self.assertEqual(2, qs.count())
+        names = [child.name for child in qs]
+        self.assertEqual(['Jan', 'Jan'], names)
+
+    def test_should_find_child_with_same_last_name(self):
+        url = reverse("admin:childlist")
+        data = {"last_name": "Pietraszek"}
+        response = self.client.get(url, data, follow=True)
```
Comment (rsiera) [inline tree/tests.py]: The same

File: tree/tests.py [hunk 9]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+
+        qs = response.context_data['childs']
+        self.assertEqual(2, qs.count())
+        last_names = [child.last_name for child in qs]
+        self.assertEqual(['Pietraszek', 'Pietraszek'], last_names)
+
+    def test_should_find_child_wtih_same_birth(self):
```
Comment (rsiera) [inline tree/tests.py]: with

File: tree/tests.py [hunk 10]
```diff
@@ -56,25 +61,149 @@ def test_i_am_not_in_the_first_place_in_database(self):
...
+        url = reverse("admin:tree_father_changelist")
+        father = Father.objects.get(id=2)
+        father.email = "marcin@pieraszek.pl"
+        data = {"action": "send_email", "_selected_action": [father.pk]}
+        response = self.client.post(url, data, follow=True)
+
+        self.assertTrue(mock_sendmail.called)
+
+    @patch("tree.admin.Thread")
+    def test_should_send_multiple_mail(self, mock_sendmail):
```
Comment (rsiera) [inline tree/tests.py]: it's not mock_sendmail, you're mocking Thread instance, so you don't check is send_mail target function

**Comment by rsiera** [comment]
Don't ask questions generally like, let's do PR and add comment - here I'm expecting POST, but method is GET, why ?

When I override modeladmin - changelist_view on get_queryset method. How I can go inside this queryset in unittesting - which ModelAdmin ? what does it mean 'go inside' ?

Why in unittesting if I write self.c.post(....) when debuging and dig into response - method is GET. - I'm guessing that because you're using follow=True.

In unittesting if I compare queryset from response.context_data (all children) with model (for example Child.objects.all() ) it is not equal. - qs from response are not ordered, you can do self.assertEqual(list(qs), list(expected_qs)
or self.assertQuerysetEqual(,,ordered=False) or sort them or many ways to handle that

**Comment by rsiera** [comment]
Wrong way, you're overcomplicating things. Threading is used with sockets when you're doing multi-threaded server, why are you start learning from multi-threaded not single-threaded environment. Just don't use Thread at this moment.

**Review by rsiera — APPROVED** [review]
Approving because of the changes in #6 PR

**Pull request merged**